File: windows/Anchor/ViewModels/AnchorViewModel.cs
using System.Collections.ObjectModel;
using System.Text.Json;
using Anchor.Data;
using Anchor.Models;
using CommunityToolkit.Mvvm.ComponentModel;

namespace Anchor.ViewModels;

public partial class PurchaseFilters : ObservableObject
{
    [ObservableProperty] private string search = "";
    [ObservableProperty] private string? category;
    [ObservableProperty] private string? supplierId;
    [ObservableProperty] private string? status;
}

public partial class AnchorViewModel : ObservableObject
{
    private readonly Store _store;

    [ObservableProperty] private AppData data;
    [ObservableProperty] private bool locked = true;
    [ObservableProperty] private string pinEntry = "";
    [ObservableProperty] private string pinError = "";
    [ObservableProperty] private string range = "all";
    [ObservableProperty] private string toast = "";
    [ObservableProperty] private bool showGreet = false;

    public PurchaseFilters Filters { get; } = new();

    public string GreetText => $"Welcome back, {Data.Settings.OwnerName} — Your wife loves you ♥";

    private string? _pinFirstEntry;
    private DateTime _lastActivity = DateTime.Now;
    private readonly Microsoft.UI.Dispatching.DispatcherQueueTimer _greetTimer;

    public AnchorViewModel(Store store)
    {
        _store = store;
        data = _store.Load();
        _greetTimer = Microsoft.UI.Dispatching.DispatcherQueue.GetForCurrentThread().CreateTimer();
        _greetTimer.Interval = TimeSpan.FromSeconds(8);
        _greetTimer.IsRepeating = false;
        _greetTimer.Tick += (_, _) => ShowGreet = false;
    }

    private void TriggerGreetIfNamed()
    {
        if (string.IsNullOrWhiteSpace(Data.Settings.OwnerName)) return;
        ShowGreet = true;
        _greetTimer.Stop();
        _greetTimer.Start();
    }

    private void Persist() => _store.Save(Data);

    public string SupplierName(string id) => Data.Suppliers.FirstOrDefault(s => s.Id == id)?.Name ?? id;

    public string Money(double? n, int? decimals = null) => Format.Money(n, Data.Settings.Currency, decimals);

    public void ShowToast(string msg) => Toast = msg;
    public void ClearToast() => Toast = "";

    public void RecordActivity() => _lastActivity = DateTime.Now;

    public void CheckAutoLock()
    {
        var minutes = Data.Settings.AutoLockMinutes;
        if (minutes > 0 && !Locked && (DateTime.Now - _lastActivity).TotalMinutes > minutes)
        {
            Locked = true;
            PinEntry = "";
        }
    }

    public void LockNow()
    {
        Locked = true;
        PinEntry = "";
    }

    public string PinMode() => Data.Settings.Pin == null ? (_pinFirstEntry == null ? "create" : "confirm") : "enter";

    public void PressPinKey(string key)
    {
        var mode = PinMode();
        var entry = key == "⌫" ? (PinEntry.Length > 0 ? PinEntry[..^1] : PinEntry) : (PinEntry + key);
        if (entry.Length > 4) entry = entry[..4];

        if (entry.Length == 4)
        {
            switch (mode)
            {
                case "create":
                    _pinFirstEntry = entry;
                    PinEntry = "";
                    PinError = "";
                    break;
                case "confirm":
                    if (entry == _pinFirstEntry)
                    {
                        Data.Settings.Pin = entry;
                        Persist();
                        Locked = false;
                        PinEntry = "";
                        _pinFirstEntry = null;
                        PinError = "";
                        TriggerGreetIfNamed();
                    }
                    else
                    {
                        PinEntry = "";
                        _pinFirstEntry = null;
                        PinError = "PINs did not match — start over";
                    }
                    break;
                default:
                    if (entry == Data.Settings.Pin)
                    {
                        Locked = false;
                        PinEntry = "";
                        PinError = "";
                        RecordActivity();
                        TriggerGreetIfNamed();
                    }
                    else
                    {
                        PinEntry = "";
                        PinError = "Wrong PIN — try again";
                    }
                    break;
            }
        }
        else
        {
            PinEntry = entry;
            PinError = "";
        }
    }

    // ----- filters / range -----
    public void SetRange(string r) => Range = r;

    public void SetStatusFilter(string? status)
    {
        Filters.Status = status;
        Filters.Category = null;
        Filters.SupplierId = null;
    }

    public void ClearFilters()
    {
        Filters.Search = "";
        Filters.Category = null;
        Filters.SupplierId = null;
        Filters.Status = null;
    }

    public List<Purchase> RangedPurchases() => Logic.RangedPurchases(Data.Purchases, Range);

    public List<Purchase> FilteredPurchases() => Logic.FilteredPurchases(
        Data.Purchases, SupplierName, Filters.Search, Filters.Category, Filters.SupplierId, Filters.Status);

    // ----- purchases -----
    public void AddPurchase(string item, string category, string supplierName, double qty, string unit, double price, string date, string status, string notes, string attach)
    {
        if (string.IsNullOrWhiteSpace(item)) { ShowToast("Item name is required."); return; }
        if (qty <= 0) { ShowToast("Quantity must be greater than zero."); return; }
        if (price <= 0) { ShowToast("Unit price must be greater than zero."); return; }
        if (string.IsNullOrWhiteSpace(supplierName)) { ShowToast("Supplier name is required."); return; }

        var supplier = FindOrCreateSupplier(supplierName.Trim());
        var id = "p" + DateTimeOffset.Now.ToUnixTimeMilliseconds();
        var docs = attach == "None" ? new List<Doc>() : new List<Doc> { new() { Name = $"{attach}-{item.Split(' ')[0]}.pdf", Type = attach, Date = date } };
        Data.Purchases.Add(new Purchase
        {
            Id = id, Item = item.Trim(), Category = category, SupplierId = supplier.Id, Qty = qty, Unit = unit,
            Price = price, Date = date, Status = status, Notes = notes.Trim(), Basis = null, GroupId = null, Docs = docs,
        });
        Persist();
        OnPropertyChanged(nameof(Data));
        ShowToast($"Purchase logged — {Money(qty * price, 0)}");
    }

    public void DeletePurchase(string id)
    {
        Data.Purchases.RemoveAll(p => p.Id == id);
        Data.Reminders.RemoveAll(r => r.LinkType == "p" && r.LinkId == id);
        foreach (var g in Data.Groups) if (g.PurchaseId == id) g.PurchaseId = null;
        Persist();
        OnPropertyChanged(nameof(Data));
        ShowToast("Purchase deleted");
    }

    private Supplier FindOrCreateSupplier(string name)
    {
        var existing = Data.Suppliers.FirstOrDefault(s => string.Equals(s.Name, name, StringComparison.OrdinalIgnoreCase));
        if (existing != null) return existing;
        var fresh = new Supplier { Id = "s" + DateTimeOffset.Now.ToUnixTimeMilliseconds(), Name = name, Contact = "", Notes = "" };
        Data.Suppliers.Add(fresh);
        return fresh;
    }

    // ----- quotes -----
    public void AddQuoteToGroup(string? groupTitle, string newTitle, string category, string supplierName, double qty, double price, string validUntil)
    {
        if (groupTitle == null && string.IsNullOrWhiteSpace(newTitle)) { ShowToast("Name the new comparison."); return; }
        if (qty <= 0 || price <= 0) { ShowToast("Quantity and unit price must be greater than zero."); return; }
        if (string.IsNullOrWhiteSpace(supplierName)) { ShowToast("Supplier name is required."); return; }

        var supplier = FindOrCreateSupplier(supplierName.Trim());
        QuoteGroup group;
        if (groupTitle == null)
        {
            group = new QuoteGroup { Id = "g" + DateTimeOffset.Now.ToUnixTimeMilliseconds(), Title = newTitle.Trim(), Qty = qty, Unit = "pcs", Category = category, Status = "active" };
            Data.Groups.Add(group);
        }
        else
        {
            group = Data.Groups.First(g => g.Title == groupTitle);
        }
        group.Quotes.Add(new Quote { Id = "q" + DateTimeOffset.Now.ToUnixTimeMilliseconds(), GroupId = group.Id, SupplierId = supplier.Id, Price = price, ValidUntil = string.IsNullOrWhiteSpace(validUntil) ? "2026-07-31" : validUntil, Status = "Received" });
        Persist();
        OnPropertyChanged(nameof(Data));
        ShowToast("Quote added");
    }

    public void SelectQuote(string groupId, string quoteId)
    {
        var group = Data.Groups.First(g => g.Id == groupId);
        group.SelectedQuoteId = quoteId;
        foreach (var q in group.Quotes) q.Status = q.Id == quoteId ? "Selected" : "Rejected";
        Persist();
        OnPropertyChanged(nameof(Data));
        ShowToast("Quote selected — savings locked in");
    }

    public void ConvertGroupToPurchase(string groupId)
    {
        var group = Data.Groups.First(g => g.Id == groupId);
        var quote = group.Quotes.First(q => q.Id == group.SelectedQuoteId);
        var highest = group.Quotes.Max(q => q.Price) * group.Qty;
        var id = "p" + DateTimeOffset.Now.ToUnixTimeMilliseconds();
        Data.Purchases.Add(new Purchase
        {
            Id = id, Item = group.Title, Category = group.Category, SupplierId = quote.SupplierId, Qty = group.Qty, Unit = group.Unit,
            Price = quote.Price, Date = Format.Today(), Status = "Ordered", Notes = "Created from quote comparison.", Basis = highest, GroupId = groupId,
            Docs = new List<Doc> { new() { Name = $"Quote-{SupplierName(quote.SupplierId).Split(' ')[0]}.pdf", Type = "Quote", Date = quote.ValidUntil } },
        });
        group.Status = "done";
        group.PurchaseId = id;
        Persist();
        OnPropertyChanged(nameof(Data));
        ShowToast("Converted to purchase");
    }

    // ----- budgets -----
    public void SetBudget(string category, double amount)
    {
        var b = Data.Budgets.FirstOrDefault(x => x.Category == category);
        if (b != null) b.Amount = amount;
        Persist();
        OnPropertyChanged(nameof(Data));
    }

    // ----- reminders -----
    public void ToggleReminderDone(string reminderId)
    {
        var r = Data.Reminders.FirstOrDefault(x => x.Id == reminderId);
        if (r == null) return;
        r.Status = r.Status == "Completed" ? "Upcoming" : "Completed";
        Persist();
        OnPropertyChanged(nameof(Data));
    }

    // ----- scan (demo) -----
    public void SaveScannedPurchase(string item, string category, string supplierName, double qty, double price, string date, string invNo, string tax)
    {
        if (string.IsNullOrWhiteSpace(item)) { ShowToast("Line item is required."); return; }
        if (qty <= 0 || price <= 0) { ShowToast("Quantity and unit price must be greater than zero."); return; }
        if (string.IsNullOrWhiteSpace(supplierName)) { ShowToast("Supplier name is required."); return; }

        var supplier = FindOrCreateSupplier(supplierName.Trim());
        var id = "p" + DateTimeOffset.Now.ToUnixTimeMilliseconds();
        var notes = !string.IsNullOrWhiteSpace(invNo)
            ? $"Scanned from {invNo}" + (!string.IsNullOrWhiteSpace(tax) && double.TryParse(tax, out var taxVal) ? $" · tax {Money(taxVal)}" : "")
            : "Scanned document";
        Data.Purchases.Add(new Purchase
        {
            Id = id, Item = item.Trim(), Category = string.IsNullOrWhiteSpace(category) ? "Office Supplies" : category, SupplierId = supplier.Id,
            Qty = qty, Unit = "pcs", Price = price, Date = date, Status = "Paid", Notes = notes,
            Docs = new List<Doc> { new() { Name = $"{(string.IsNullOrWhiteSpace(invNo) ? "scan" : invNo)}.pdf", Type = "Invoice", Date = date } },
        });
        Persist();
        OnPropertyChanged(nameof(Data));
        ShowToast("Scanned purchase saved");
    }

    public string? DupCheckForScan(string scannedSupplierName, double qty, double price, string date) =>
        Logic.DupCheck(Data.Purchases, SupplierName, scannedSupplierName, qty, price, date, Data.Settings.Currency);

    // ----- settings -----
    public void SaveSettings(string currency, bool appLockEnabled, int budgetAlertPct, int autoLockMinutes, string ownerName)
    {
        Data.Settings.Currency = currency;
        Data.Settings.AppLockEnabled = appLockEnabled;
        Data.Settings.BudgetAlertPct = budgetAlertPct;
        Data.Settings.AutoLockMinutes = autoLockMinutes;
        Data.Settings.OwnerName = ownerName;
        Persist();
        OnPropertyChanged(nameof(Data));
    }

    public string ExportBackupJson() => Backup.Export(Data);

    public string RestoreBackup(string json)
    {
        var result = Backup.Restore(json);
        if (result == null) return "Restore failed — file could not be read as a backup. Your current data is untouched.";
        Data = result;
        Persist();
        OnPropertyChanged(nameof(Data));
        return $"Backup restored — {result.Purchases.Count} purchases";
    }

    public void ResetToSampleData()
    {
        var fresh = Seed.Full();
        fresh.Settings.Pin = Data.Settings.Pin;
        Data = fresh;
        Persist();
        OnPropertyChanged(nameof(Data));
        ShowToast("Sample data loaded");
    }

    public void ClearAllData()
    {
        var fresh = Seed.Blank();
        fresh.Settings = Data.Settings;
        Data = fresh;
        Persist();
        OnPropertyChanged(nameof(Data));
        ShowToast("All data cleared — fresh start");
    }
}

File: android/app/src/main/kotlin/com/anchor/procurement/ui/AnchorViewModel.kt
package com.anchor.procurement.ui

import androidx.lifecycle.ViewModel
import androidx.lifecycle.ViewModelProvider
import androidx.lifecycle.viewModelScope
import com.anchor.procurement.data.AnchorData
import com.anchor.procurement.data.Backup
import com.anchor.procurement.data.BudgetEntity
import com.anchor.procurement.data.Format
import com.anchor.procurement.data.Logic
import com.anchor.procurement.data.Purchase
import com.anchor.procurement.data.PurchaseDocEntity
import com.anchor.procurement.data.PurchaseEntity
import com.anchor.procurement.data.QuoteEntity
import com.anchor.procurement.data.QuoteGroupEntity
import com.anchor.procurement.data.ReminderEntity
import com.anchor.procurement.data.Repository
import com.anchor.procurement.data.SettingsEntity
import kotlinx.coroutines.Job
import kotlinx.coroutines.delay
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.SharingStarted
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.flow.stateIn
import kotlinx.coroutines.launch

data class PurchaseFilters(
    val search: String = "",
    val category: String? = null,
    val supplierId: String? = null,
    val status: String? = null,
)

class AnchorViewModel(private val repository: Repository) : ViewModel() {

    val data: StateFlow<AnchorData> = repository.data.stateIn(
        viewModelScope, SharingStarted.WhileSubscribed(5000), AnchorData(),
    )

    private val _locked = MutableStateFlow(true)
    val locked: StateFlow<Boolean> = _locked.asStateFlow()

    private val _pinEntry = MutableStateFlow("")
    val pinEntry: StateFlow<String> = _pinEntry.asStateFlow()

    private val _pinFirstEntry = MutableStateFlow<String?>(null)
    private val _pinError = MutableStateFlow("")
    val pinError: StateFlow<String> = _pinError.asStateFlow()

    private val _range = MutableStateFlow("all")
    val range: StateFlow<String> = _range.asStateFlow()

    private val _filters = MutableStateFlow(PurchaseFilters())
    val filters: StateFlow<PurchaseFilters> = _filters.asStateFlow()

    private val _toast = MutableStateFlow("")
    val toast: StateFlow<String> = _toast.asStateFlow()

    private val _showGreet = MutableStateFlow(false)
    val showGreet: StateFlow<Boolean> = _showGreet.asStateFlow()
    private var greetJob: Job? = null

    val greetText: String
        get() = "Welcome back, ${data.value.settings.ownerName} — Your wife loves you ♥"

    private var lastActivityAt = System.currentTimeMillis()

    private fun triggerGreetIfNamed() {
        if (data.value.settings.ownerName.isBlank()) return
        _showGreet.value = true
        greetJob?.cancel()
        greetJob = viewModelScope.launch {
            delay(8000)
            _showGreet.value = false
        }
    }

    fun supplierName(id: String): String = data.value.suppliers.find { it.id == id }?.name ?: id

    fun money(n: Double?, decimals: Int? = null) = Format.money(n, data.value.settings.currency, decimals)

    fun showToast(msg: String) {
        _toast.value = msg
    }

    fun clearToast() {
        _toast.value = ""
    }

    fun recordActivity() {
        lastActivityAt = System.currentTimeMillis()
    }

    fun checkAutoLock() {
        val minutes = data.value.settings.autoLockMinutes
        if (minutes > 0 && !_locked.value && System.currentTimeMillis() - lastActivityAt > minutes * 60_000L) {
            _locked.value = true
            _pinEntry.value = ""
        }
    }

    fun lockNow() {
        _locked.value = true
        _pinEntry.value = ""
    }

    /** Clears the saved PIN (data stays intact) so the lock screen re-enters "create" mode. */
    fun forgotPin() {
        viewModelScope.launch {
            repository.saveSettings(currentSettingsEntity().copy(pin = null))
        }
        _pinFirstEntry.value = null
        _pinEntry.value = ""
        _pinError.value = ""
    }

    /** Mirrors the prototype's pinPress: builds up to 4 digits, then creates/confirms/verifies. */
    fun pressPinKey(key: String) {
        val pin = data.value.settings.pin
        val mode = if (pin == null) (if (_pinFirstEntry.value == null) "create" else "confirm") else "enter"
        var entry = _pinEntry.value
        entry = if (key == "⌫") entry.dropLast(1) else (entry + key).take(4)

        if (entry.length == 4) {
            when (mode) {
                "create" -> {
                    _pinFirstEntry.value = entry
                    _pinEntry.value = ""
                    _pinError.value = ""
                }
                "confirm" -> {
                    if (entry == _pinFirstEntry.value) {
                        viewModelScope.launch {
                            repository.saveSettings(currentSettingsEntity().copy(pin = entry))
                        }
                        _locked.value = false
                        _pinEntry.value = ""
                        _pinFirstEntry.value = null
                        _pinError.value = ""
                        triggerGreetIfNamed()
                    } else {
                        _pinEntry.value = ""
                        _pinFirstEntry.value = null
                        _pinError.value = "PINs did not match — start over"
                    }
                }
                else -> {
                    if (entry == pin) {
                        _locked.value = false
                        _pinEntry.value = ""
                        _pinError.value = ""
                        recordActivity()
                        triggerGreetIfNamed()
                    } else {
                        _pinEntry.value = ""
                        _pinError.value = "Wrong PIN — try again"
                    }
                }
            }
        } else {
            _pinEntry.value = entry
            _pinError.value = ""
        }
    }

    fun pinMode(): String {
        val pin = data.value.settings.pin
        return if (pin == null) (if (_pinFirstEntry.value == null) "create" else "confirm") else "enter"
    }

    private fun currentSettingsEntity(): SettingsEntity {
        val s = data.value.settings
        return SettingsEntity(0, s.pin, s.autoLockMinutes, s.appLockEnabled, s.currency, s.budgetAlertPct, s.ownerName)
    }

    // ----- filters / range -----
    fun setRange(r: String) { _range.value = r }
    fun setSearch(q: String) { _filters.value = _filters.value.copy(search = q) }
    fun setStatusFilter(status: String?) { _filters.value = _filters.value.copy(status = status, category = null, supplierId = null) }
    fun setCategoryFilter(category: String?) { _filters.value = _filters.value.copy(category = category) }
    fun setSupplierFilter(supplierId: String?) { _filters.value = _filters.value.copy(supplierId = supplierId) }
    fun clearFilters() { _filters.value = PurchaseFilters() }

    fun rangedPurchases(): List<Purchase> = Logic.rangedPurchases(data.value.purchases, _range.value)
    fun filteredPurchases(): List<Purchase> {
        val f = _filters.value
        return Logic.filteredPurchases(data.value.purchases, ::supplierName, f.search, f.category, f.supplierId, f.status)
    }

    // ----- purchases -----
    fun addPurchase(
        item: String, category: String, supplierName: String, qty: Double, unit: String,
        price: Double, date: String, status: String, notes: String, attach: String,
    ) {
        if (item.isBlank()) { showToast("Item name is required."); return }
        if (qty <= 0) { showToast("Quantity must be greater than zero."); return }
        if (price <= 0) { showToast("Unit price must be greater than zero."); return }
        if (supplierName.isBlank()) { showToast("Supplier name is required."); return }
        viewModelScope.launch {
            val supplier = repository.findOrCreateSupplier(supplierName.trim())
            val id = "p${System.currentTimeMillis()}"
            val docs = if (attach == "None") emptyList() else listOf(
                PurchaseDocEntity(purchaseId = id, name = "$attach-${item.split(" ").first()}.pdf", type = attach, date = date),
            )
            repository.addPurchase(
                PurchaseEntity(id, item.trim(), category, supplier.id, qty, unit, price, date, status, notes.trim(), null, null),
                docs,
            )
            showToast("Purchase logged — ${money(qty * price, 0)}")
        }
    }

    fun deletePurchase(id: String) {
        viewModelScope.launch {
            repository.deletePurchase(id)
            showToast("Purchase deleted")
        }
    }

    // ----- quotes -----
    fun addQuoteToGroup(groupTitle: String?, newTitle: String, category: String, supplierName: String, qty: Double, price: Double, validUntil: String) {
        if (groupTitle == null && newTitle.isBlank()) { showToast("Name the new comparison."); return }
        if (qty <= 0 || price <= 0) { showToast("Quantity and unit price must be greater than zero."); return }
        if (supplierName.isBlank()) { showToast("Supplier name is required."); return }
        viewModelScope.launch {
            val supplier = repository.findOrCreateSupplier(supplierName.trim())
            val group = if (groupTitle == null) {
                val g = QuoteGroupEntity("g${System.currentTimeMillis()}", newTitle.trim(), qty, "pcs", category, "active", null, null)
                repository.addQuoteGroup(g)
                g
            } else {
                data.value.groups.first { it.title == groupTitle }.let {
                    QuoteGroupEntity(it.id, it.title, it.qty, it.unit, it.category, it.status, it.selectedQuoteId, it.purchaseId)
                }
            }
            repository.addQuote(QuoteEntity("q${System.currentTimeMillis()}", group.id, supplier.id, price, validUntil.ifBlank { "2026-07-31" }, "Received"))
            showToast("Quote added")
        }
    }

    fun selectQuote(groupId: String, quoteId: String) {
        viewModelScope.launch {
            val group = data.value.groups.first { it.id == groupId }
            repository.updateQuoteGroup(
                QuoteGroupEntity(group.id, group.title, group.qty, group.unit, group.category, group.status, quoteId, group.purchaseId),
            )
            group.quotes.forEach { q ->
                repository.updateQuote(QuoteEntity(q.id, q.groupId, q.supplierId, q.price, q.validUntil, if (q.id == quoteId) "Selected" else "Rejected"))
            }
            showToast("Quote selected — savings locked in")
        }
    }

    fun convertGroupToPurchase(groupId: String) {
        viewModelScope.launch {
            val group = data.value.groups.first { it.id == groupId }
            val quote = group.quotes.first { it.id == group.selectedQuoteId }
            val highest = (group.quotes.maxOfOrNull { it.price } ?: 0.0) * group.qty
            val id = "p${System.currentTimeMillis()}"
            repository.addPurchase(
                PurchaseEntity(
                    id, group.title, group.category, quote.supplierId, group.qty, group.unit,
                    quote.price, Format.today(), "Ordered", "Created from quote comparison.", highest, groupId,
                ),
                listOf(PurchaseDocEntity(purchaseId = id, name = "Quote-${supplierName(quote.supplierId).split(" ").first()}.pdf", type = "Quote", date = quote.validUntil)),
            )
            repository.updateQuoteGroup(QuoteGroupEntity(group.id, group.title, group.qty, group.unit, group.category, "done", group.selectedQuoteId, id))
            showToast("Converted to purchase")
        }
    }

    // ----- budgets -----
    fun setBudget(category: String, amount: Double) {
        viewModelScope.launch { repository.setBudgetAmount(category, amount) }
    }

    // ----- reminders -----
    fun toggleReminderDone(reminderId: String) {
        viewModelScope.launch {
            val r = data.value.reminders.first { it.id == reminderId }
            val newStatus = if (r.status == "Completed") "Upcoming" else "Completed"
            repository.toggleReminder(ReminderEntity(r.id, r.title, r.linkType, r.linkId, r.due, newStatus))
        }
    }

    // ----- scan (demo) -----
    fun saveScannedPurchase(item: String, category: String, supplierName: String, qty: Double, price: Double, date: String, invNo: String, tax: String) {
        if (item.isBlank()) { showToast("Line item is required."); return }
        if (qty <= 0 || price <= 0) { showToast("Quantity and unit price must be greater than zero."); return }
        if (supplierName.isBlank()) { showToast("Supplier name is required."); return }
        viewModelScope.launch {
            val supplier = repository.findOrCreateSupplier(supplierName.trim())
            val id = "p${System.currentTimeMillis()}"
            val notes = if (invNo.isNotBlank()) "Scanned from $invNo" + (if (tax.isNotBlank()) " · tax ${money(tax.toDoubleOrNull())}" else "") else "Scanned document"
            repository.addPurchase(
                PurchaseEntity(id, item.trim(), category.ifBlank { "Office Supplies" }, supplier.id, qty, "pcs", price, date, "Paid", notes, null, null),
                listOf(PurchaseDocEntity(purchaseId = id, name = "${invNo.ifBlank { "scan" }}.pdf", type = "Invoice", date = date)),
            )
            showToast("Scanned purchase saved")
        }
    }

    fun dupCheckForScan(scannedSupplierName: String, qty: Double, price: Double, date: String): String? =
        Logic.dupCheck(data.value.purchases, ::supplierName, scannedSupplierName, qty, price, date, data.value.settings.currency)

    // ----- settings -----
    fun saveSettings(currency: String, appLockEnabled: Boolean, budgetAlertPct: Int, autoLockMinutes: Int, ownerName: String) {
        viewModelScope.launch {
            repository.saveSettings(currentSettingsEntity().copy(currency = currency, appLockEnabled = appLockEnabled, budgetAlertPct = budgetAlertPct, autoLockMinutes = autoLockMinutes, ownerName = ownerName))
        }
    }

    fun setAutoLockMinutes(minutes: Int) {
        viewModelScope.launch { repository.saveSettings(currentSettingsEntity().copy(autoLockMinutes = minutes)) }
    }

    fun exportBackupJson(): String = Backup.export(data.value)

    fun restoreBackup(json: String, onDone: (String) -> Unit) {
        viewModelScope.launch {
            repository.restoreFromBackup(json).fold(
                onSuccess = { count -> onDone("Backup restored — $count purchases") },
                onFailure = { onDone("Restore failed — file could not be read as a backup. Your current data is untouched.") },
            )
        }
    }

    fun resetToSampleData() {
        viewModelScope.launch {
            repository.loadSampleData()
            showToast("Sample data loaded")
        }
    }

    fun clearAllData() {
        viewModelScope.launch {
            repository.clearAllData()
            showToast("All data cleared — fresh start")
        }
    }

    class Factory(private val repository: Repository) : ViewModelProvider.Factory {
        override fun <T : ViewModel> create(modelClass: Class<T>): T {
            @Suppress("UNCHECKED_CAST")
            return AnchorViewModel(repository) as T
        }
    }
}
